### PR TITLE
ar/projects with many files

### PR DESCRIPTION
**Pull request by anselrognlie**
re-work for projects with many files

- the revised structure is still largely inherited from the original, but I strongly felt that a more detailed look at the most common import methods would help anchor the discussion
- also introduces __main__.py as am optional place to put startup logic (with parallels to pytest and venv)
- adds an additional CFU
- fleshes out existing CFU explanation
- the overall lesson tries to emphasize package setup and imports as they pertain to packages rather than more free-floating modules

### CODE REVIEW

File: intro-to-oop-contd/projects-with-many-files.md [hunk 1]
```diff
@@ -147,40 +171,176 @@ We must _import_ any resources that we need from another module before we can us
...
+Hopefully good naming practices were employed to help hint at the proper usage of different identifiers. But if we import an identifier that refers to a value (a variable) but then try to call it like a function, this will result in a runtime error! So it's our job to read and understand the code or documentation of the module we want to use so that we import and make use of its identifiers correctly!
 
 ### !end-callout
 
-### Example Between Classes
+#### 3. Importing Module Identifiers using Relative Names
+
+Importing package modules by full name works everywhere, whether from outside or inside a package. But to access the identifiers defined in the module, we have to prefix them with the full package name. If our package were several layers deep, this could make for some long names!
+
+Let's say that the `passenger` module in our `ride_share_app` package needs to make use of the `Driver` class. We could certainly import the `driver` module as before and access the `Driver` class with the full package name. But `driver.py` and `passenger.py` are defined in the same package. In fact, they're in the same folder. Wouldn't it be nice if there were a shorter way?
```
Comment (tildeee) [inline intro-to-oop-contd/projects-with-many-files.md:230]: I know it's a lot of copy/pasting, but I think that copy/pasting and repeating the `ride_share_app` structure before this paragraph would help visualize to students that `driver.py` and `passenger.py` are in the same folder.

File: intro-to-oop-contd/projects-with-many-files.md [hunk 2]
```diff
@@ -147,40 +171,176 @@ We must _import_ any resources that we need from another module before we can us
...
+    ├── __init__.py
+    ├── driver.py
+    ├── passenger.py
+    └── payment
+        ├── __init__.py
+        ├── cash.py
+        └── credit.py
+```
+
+If our package had a folder called `payment` under `ride_share_app` (with an `__init__.py` file), and `payment` had the modules `cash.py` and `credit.py`, then from either of these modules we could access the `Driver` identifier in `driver.py` as:
```
Comment (tildeee) [inline intro-to-oop-contd/projects-with-many-files.md:266]: My preference is to think of this `payment` section as a second paragraph.

With that, I think I'd recommend:

- Moving this line to directly above the example tree structure (currently directly above this paragraph)
- Modifying the wording of this line to introduce this as a second example
- Frame this second example to solving a different question: What's the syntax for importing something from two folders up?
- Name the modules `cash.py` and `credit.py` in a simplified sense instead of saying "then from either of these modules." (take a more concrete stance, so a student brain doesn't have to parse through "is it for cash.py or credit.py, what is a module, which folder are we looking at," etc)

```suggestion
Let's consider a second, similar situation. How do we import a module that is inside a different folder, but within the same package?

Imagine that our package had a folder called `payment` under `ride_share_app` (with an `__init__.py` file), and `payment` had the modules `cash.py` and `credit.py`. Inside either `cash.py` or `credit.py`, we can access the `Driver` identifier in `driver.py` as:
```

File: intro-to-oop-contd/projects-with-many-files.md [hunk 3]
```diff
@@ -147,40 +171,176 @@ We must _import_ any resources that we need from another module before we can us
...
+
+## Be Careful With Descriptive Entry Point File Names
+
+Scarlet also briefly considered renaming `main.py` to `ride_share_app.py`, but then she realized that would prevent Python from being able to see the `ride_share_app` package. If the main file has the same name as the package, if we try to import from the package, Python will try looking inside the module we are currently in, not find the desired identifier, and raise an `ImportError`.
+
+<br />
+
+We should make sure that the entry point file name does _not_ match the project package name. Alternatives include adding `_pkg` to the name of the package, though this isn't very Pythonic. Or we could make the entry point closely related to, but not the same as the package name, like how Scarlet considered `ride_share.py`. We can use a more general entry point, like `main.py`. Or if we want to be able to use the package name to launch the application, we can use the special `__main__.py` file described later in this lesson.
+
+### !end-callout
```
Comment (tildeee) [inline intro-to-oop-contd/projects-with-many-files.md:335]: The students don't necessarily have meaningful context or examples to have run into this, felt this pain, or the motivation to do something with this knowledge, IMO.

I think it personally takes until capstone to begin to have opinions on file names.

Also, the concept of an "entry point" is pretty intangible and far to them IMO. The closest entry point that they may see may be a Flask project depending on how you think about it. Even at that point, I think it's a lot of explanation to think about an entry point and it'd get awkward without the proper way to visualize it.

In that whole theme, I think that the idea of Scarlet considering renaming `main.py` is a really out-of-reach consideration. I don't think the knowledge/scenario presented here will be the juiciest for students.

Personally, I'm into the idea of cutting it, to make room for the other good content in here?

```suggestion
```

File: intro-to-oop-contd/projects-with-many-files.md [hunk 4]
```diff
@@ -147,40 +171,176 @@ We must _import_ any resources that we need from another module before we can us
...
+two.do_something_else()
+```
+
+### So Many Ways to Import
+
+As we have shown, there are _many_ different syntaxes and strategies we can use to import modules. If a particular import syntax works for a project, it's valid!
+
+Different situations will need to use different ways of importing. So we should be ready to see many styles of importing. We may even encounter styles that we haven't discussed that we might need to look up. That's ok! But we should feel confident that we can recognize importing when we see it!
+
+## Restructuring a Project
```
Comment (tildeee) [inline intro-to-oop-contd/projects-with-many-files.md:317]: Soft suggestion/question? I found myself craving that this section was framed as an example walkthrough. Your first sentence in this section pretty much states that though. Would it be valuable to rename this to something like:

```suggestion
## Example: Restructuring a Project
```

```suggestion
## Example Walkthrough: Restructuring a Project
```

or something to that effect?

File: intro-to-oop-contd/projects-with-many-files.md [hunk 5]
```diff
@@ -264,79 +501,301 @@ Where `passenger.py` looks like:
...
 
 * type: multiple-choice
 * id: f032adcb-c05b-4591-96b9-58bc24d7ee79
-* title: Importing modules
+* title: Projects With Many Files
 
 ##### !question
 
-Hosts for AdaBnB can host guests at multiple properties on the app. Each property is an instance of a `Place` class. If we ran `guest.py`, we would receive a `NameError`. Select the option that describes how to solve this `NameError`.
+Hosts for AdaBnB can host guests at multiple properties on the app. Each property is an instance of a `Place` class. If we tried to call the `create_places` method of `Host`, we would receive a `NameError`. Assuming the `place` module in which the `Place` class is defined is in the same package folder as `host.py`, select the option that best describes how to solve this `NameError`.
```
Comment (tildeee) [inline intro-to-oop-contd/projects-with-many-files.md:756]: Soft suggestion: The sentence is good, but it might be nice to give them a diagram either additonally, or instead.

For convenience

```
adabnb/
├── README.md
├── requirements.txt
├── adabnb
│   ├── __init__.py
│   ├── host.py
│   └── place.py
└── tests
    ├── __init__.py
    ├── host_test.py
    └── place_test.py
```

Didn't make a suggestion _because code formatting_.

**Review by tildeee — CHANGES_REQUESTED** [review]
Fantastic lesson!!!!!

I really really really like the flow of the lesson that you had: Just some good ole clear examples and explanations and options here. I really like how the three styles of importing got listed here.

Overall, my feedback is a lot around simplifying sentences and simplifying the parts that are the important bits, and not getting too deep on vocab and material. There's a bit of an undercurrent of my pulling the curriculum back towards "this is their first exposure to the material"

I wish "request changes" wasn't a big scary red; requesting changes simply as a "please check my review"

File: intro-to-oop-contd/projects-with-many-files.md [hunk 6]
```diff
@@ -33,28 +33,30 @@ Let's take a look at what we need to do to use this method of organization in ou
...
-| Package | A collection of modules                                                                | -        | "The methods and classes accessible in a package can be found in the `__init__.py` file |
+| Vocab   | Definition                                                                             | How to Use in a Sentence                                                                                                                  |
+| ------- | -------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
+| Module  | Any `.py` file that contains functions, classes, variables, and/or other runnable code | "I used the `math.Euler()` from the `Math` module to calculate an angle"                                                                  |
+| Package | A collection of modules                                                                | "The `requests` package contains modules to help us make HTTP requests." "My package has custom configuration in its `__init__.py` file." |
```
[the remark this answers is not in the document]
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:39]:
```suggestion
| Package | A collection of modules                                                                | "The `requests` package contains modules to help us request data from the Internet." "My package has custom configuration in its `__init__.py` file." |
```

[the remark this answers is not in the document]
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:39]: Not motivated enough to think of a different package, but I can rephrase it to something more easy to identify with.

File: intro-to-oop-contd/projects-with-many-files.md [hunk 7]
```diff
@@ -73,17 +75,25 @@ Often, files and folders are in the project root because it makes the most logic
 
 ### Project Code Lives in a Package Folder
 
-Projects that have classes will have a `.py` file for each class. All of these files should live in one single folder. This folder may also contain an `__init__.py` file. Details about this special file are below.
+Under the project root, we will have a single folder to contain all of our project classes, one class per `.py` file. We refer to this folder as the **package folder**. At the start of a project, when we don't have many class files, we might put them all directly in the package folder. But as the project becomes more complex, we can add subfolders to further group together related class files.
```
[the remark this answers is not in the document]
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:78]: I'll take this for now as a reminder to address this while reworking this part to get rid of the __main__ stuff. I might be able to bring the passage that used to follow this back up as part of the rearrangement, making this line specifically redundant, but it'll get addressed one way or the other.

File: intro-to-oop-contd/projects-with-many-files.md [hunk 8]
```diff
@@ -73,17 +75,25 @@ Often, files and folders are in the project root because it makes the most logic
...
+There are also a number of special files that can be placed in the package folder.
 
-The above example names this folder `project_package_name`. It contains two class files, `example_class_a.py` and `example_class_b.py`, and an `__init__.py` file.
+The most common of these is called the `__init__.py` file, which is discussed briefly in the Package Details section below.
+
+Projects with a clear starting point might include a `__main__.py` as a fairly standard place to put the application start up logic. Some projects may not need this file, and some might choose to put that logic somewhere else.
+
+Our package folder usually has the same name as the project, but it must follow some package naming rules which we will review a little later in this lesson.
+
+The above example names this folder `project_package_name` so that we can distinguish it from the project root, but there would be nothing preventing us from calling it `project_name` to match the project root name. It contains two class files, `example_class_a.py` and `example_class_b.py`, and both an `__init__.py` file and a `__main__.py`.
```
[the remark this answers is not in the document]
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:88]: Just gonna leave this as is, since I need to cull the __main__ stuff anyway. This will get addressed as part of that.

File: intro-to-oop-contd/projects-with-many-files.md [hunk 9]
```diff
@@ -147,40 +171,176 @@ We must _import_ any resources that we need from another module before we can us
...
+
+As an example, consider the `random` module, which provides the `randint` function.
 
 ```python
-from package_name.module_name import ExampleClassName
+import random
 ```
 
-It may be helpful to read this line of code in "reverse order," as "Import an object named `ExampleClassName` that's defined in the `module_name` module that's located in the `package_name` package."
+This statement imports the `random` module name into the current scope, allowing us to access identifiers in the module by using dot notation to look inside the module name as follows:
```
[the remark this answers is not in the document]
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:188]: 'This statement' refers back to the previous statement. I'd still like a for example leading into the next.

File: intro-to-oop-contd/projects-with-many-files.md [hunk 10]
```diff
@@ -147,40 +171,176 @@ We must _import_ any resources that we need from another module before we can us
...
 
-## Modules Know They're in the Same Package
+## Why We Keep Saying Identifier
 
-Sometimes, it may be unnecessary to specify a package if both modules are in the same package already. For details on the locations that Python searches for modules (and their priority order), we can read about [Python's module search path order](https://docs.python.org/3/tutorial/modules.html#the-module-search-path).
+Notice that we have been careful to say that Python imports identifiers into the current scope. Identifiers are names of things. Variables, functions, and classes are all named with identifiers. From just looking at an identifier, we don't know what it refers to. So we have to be careful to import the correct identifiers, and then use them as they are intended to be used.
+
+<br />
+
+Hopefully good naming practices were employed to help hint at the proper usage of different identifiers. But if we import an identifier that refers to a value (a variable) but then try to call it like a function, this will result in a runtime error! So it's our job to read and understand the code or documentation of the module we want to use so that we import and make use of its identifiers correctly!
```
[the remark this answers is not in the document]
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:222]: I really wish we had a part two to this lesson, but I'm sure that won't happen just now. Not necessarily here, but later in the curriculum. It feels like we're just scratching the surface in so many places, and I worry that we haven't built enough of a foundation in this one lesson (which is fairly disconnected from much of the rest of the material) for students to evaluate additional information sources. I think that contributes to my desire to talk about that next little bit, and the next little bit!

File: intro-to-oop-contd/projects-with-many-files.md [hunk 11]
```diff
@@ -147,40 +171,176 @@ We must _import_ any resources that we need from another module before we can us
...
+
+```
+└── ride_share_app
+    ├── __init__.py
+    ├── driver.py
+    ├── passenger.py
+    └── payment
+        ├── __init__.py
+        ├── cash.py
+        └── credit.py
```
[the remark this answers is not in the document]
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:263]: WHAT IS THIS WITCHCRAFT?!?!?!?

I did these all MANUALLY!!!!!

It was really that bad. I just copied an existing tree layout and used area selections to move bits around.... BUT STILL!!!!!

In any case, what happened there was that I didn't want to bring down all the test directory stuff, so initially I tried inserting some ellipses, but I didn't like it. So then I just had a dangly down bit that trailed into nothing, and I didn't like that. So then I trimmed it as you see. But I guess that's weird too.

Long story short, I'll take your suggestion, then sit in a corner and sob softly.

File: intro-to-oop-contd/projects-with-many-files.md (hunk 2 above)
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:266]: Taking the suggestion. I'll move the lines around locally.

File: intro-to-oop-contd/projects-with-many-files.md (hunk 3 above)
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:335]: You already heard the back story for this. But as we're going to deemphasize the entry point stuff, this can go!

File: intro-to-oop-contd/projects-with-many-files.md (hunk 5 above)
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:756]: Good idea! And thanks for the tree!

File: intro-to-oop-contd/projects-with-many-files.md [hunk 12]
```diff
@@ -204,55 +397,117 @@ class Passenger:
...
+Notice that the `tests` folder is in the project root folder (`ride-share-app`), not in the package folder (`ride_share_app`).
+
+</details>
 
 ## Debugging Imports
 
 Here are some common bugs that we may come across with importing:
 
+### `ModuleNotFoundError`
+
```
Comment (tildeee) [inline intro-to-oop-contd/projects-with-many-files.md:507]: I could imagine a small lead-up sentence:

```suggestion
### `ModuleNotFoundError`

Consider this stack trace caused from running `passenger.py`

```

File: intro-to-oop-contd/projects-with-many-files.md [hunk 13]
```diff
@@ -264,79 +519,320 @@ Where `passenger.py` looks like:
...
 
-- A `ModuleNotFoundError` is raised because the program can't find `non_existing_module`
-- There's an import statement that can't find the specified module
-  - Double-check the `from` part. Do we need to including the package name, if this file isn't in the same package? Or do we need to exclude it?
+The `ModuleNotFoundError` is raised because the program can't find `non_existing_module`. In general, this error tells us:
+
+- There's an `import` statement that can't find the specified module
+  - Double-check the `from` part. Do we need to including the package name, if this file isn't in the same package? Ar we missing a `.` (one or more) for a relative import?
+
+### `ImportError`
```
Comment (tildeee) [inline intro-to-oop-contd/projects-with-many-files.md:528]: Similar to above

```suggestion
### `ImportError`

Consider this stack trace caused from running `passenger.py`

```

**Review by tildeee — APPROVED** [review]
AMAZING! LOOKS GREAT! feel free to check out those suggestions, but otherwise LOOKS GOOOOOOOD

File: intro-to-oop-contd/projects-with-many-files.md (hunk 12 above)
Comment (anselrognlie) [reply intro-to-oop-contd/projects-with-many-files.md:507]: Ha ha.  Yeah. This struck me as a little abrupt too, just not enough to do anything about.

Thanks for the suggestions though!  I'll take them. 🙂
